### PR TITLE
[Truffle] Lazy translation

### PR DESCRIPTION
**Pull request by chrisseaton**
Add support for lazy translation, as a first step towards lazy parsing and maybe serialisation.

It's quite effective on its own though. For `-e 14` minimum heap goes from 35 to 27 MB (saves 8 MB), total allocation from 177.85 MB ± 2.75 MB to 125.78 MB ± 2.2 MB (saves 52 MB), time from 3.47 to 3.13 s (saves 0.34 s).

We can't turn it on however, as it interferes with Truffle instrumentation. I think that's a Truffle bug and I'll be using this new lazy translation as a test case to show that. It's off by default for that reason. There might be a couple of other bugs that are ours which I'll look at after the Truffle issues is fixed.

Contains some work by Brian Belleville.

### CODE REVIEW

**Comment by chrisseaton** [comment]
Harder to measure Rails because it doesn't have an easy to observe start and end, but it seems to save around 17% for time to be up and serving.

File: truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java [hunk 1]
```diff
@@ -290,14 +293,31 @@ public RubyNode doCompileMethodBody(RubySourceSection sourceSection, String meth
     }
 
     public MethodDefinitionNode compileMethodNode(RubySourceSection sourceSection, String methodName, MethodDefParseNode defNode, ParseNode bodyNode, SharedMethodInfo sharedMethodInfo) {
-        final RubyNode body = compileMethodBody(sourceSection, methodName, bodyNode, sharedMethodInfo);
+        final TranslatorState state = getCurrentState();
```
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java]: The state only needs to be captured if we are doing lazy translation I think

**Comment by eregon** [comment]
Nice one!

File: truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        return resolve().execute(frame);
+    }
+
+    public RubyNode resolve() {
+        final RubyNode resolved = resolver.get();
+        replace(resolved, "lazy translation node resolved");
```
Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java]: Do we need to worry about two threads resolving the same method at the same time? Or maybe method aliases?

File: truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java [hunk 3]
```diff
@@ -449,4 +469,33 @@ private static String indentation(String line) {
...
 
+    /*
+     * The following methods allow us to save and restore enough of
+     * the current state of the Translator to allow lazy parsing. When
+     * the lazy parsing is actually performed, the state is restored
+     * to what it would have been if the method had been parsed
+     * eagerly.
+     */
+    public TranslatorState getCurrentState() {
+        return new TranslatorState(getEnvironment().unsafeGetLexicalScope(), getEnvironment().isDynamicConstantLookup(), new ArrayDeque<>(parentSourceSection));
```
Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java]: I don't recall the exact issue, but I had to do it in BodyTranslator a while ago (see f9df82cff1b48fb56545637de3bd1e605d19c11a). I think there was a problem with nested visit calls that each modified the `parentSourceSection`, which is just a field on the translator. Perhaps more locally scoped variables could work, but the stack approach made sense to me.

**Comment by nirvdrum** [comment]
This is surprisingly elegant.

File: truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java (hunk 3 above)
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java]: Yeah I'm just leaving the stack alone at the moment as I don't totally understand it. I'll look at it again later.

File: truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java (hunk 2 above)
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java]: Fixed.

File: truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        this.resolver = resolver;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        return resolve().execute(frame);
+    }
+
+    public RubyNode resolve() {
```
Comment (pitr-ch) [inline truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java]: could be private?

**Comment by pitr-ch** [comment]
Nice!

File: truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java (hunk 4 above)
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/parser/LazyTranslationNode.java]: I'm going to argue that Truffle should call it when it wants to instrument the node.

File: truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java (hunk 3 above)
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java]: Removed in c8e4a45e54675b8fc72a9bbab07cccc21e0fde47.

**Pull request merged**